### PR TITLE
Minor: Add more doc strings to WindowExpr

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

More work towards https://github.com/apache/arrow-datafusion/issues/5781 by incrementally adding docstrings to the relevant `WindowFunction` expressions

# Rationale for this change

This code needs to be well documented if we expose it and even if we don't, documenting it better will make the code easier to understand

# What changes are included in this PR?

More Docstrings that describe the differences of functions on `PartitionEvaluator` more

# Are these changes tested?
N/A

# Are there any user-facing changes?

### CODE REVIEW

**Review by r4ntix — APPROVED** [review]
looks like a nice improvement 👍

File: datafusion/physical-expr/src/window/built_in_window_function_expr.rs [hunk 1]
```diff
@@ -95,10 +95,13 @@ pub trait BuiltInWindowFunctionExpr: Send + Sync + std::fmt::Debug {
...
 
-    /// Does the window function use the values from its window frame?
+    /// Does the window function use the values from the window frame,
+    /// if one is specified?
     ///
     /// If this function returns true, [`Self::create_evaluator`] must
-    /// implement [`PartitionEvaluator::evaluate_inside_range`]
+    /// implement [`PartitionEvaluator::evaluate_inside_range`]. See
+    /// details and examples on [`PartitionEvaluator::evaluate`].
+
```
Comment (jackwener) [inline datafusion/physical-expr/src/window/built_in_window_function_expr.rs:104]:
```suggestion
```
redundant line

**Review by jackwener — APPROVED** [review]

File: datafusion/physical-expr/src/window/partition_evaluator.rs [hunk 2]
```diff
@@ -130,9 +130,41 @@ pub trait PartitionEvaluator: Debug + Send {
...
+    /// thus using `evaluate`, DataFusion can skip the (costly) window
+    /// frame boundary calculation.
+    ///
+    /// For example, the `LAG` built in window function does not use
+    /// the values of its window frame (it can be computed in one shot
+    /// on the entire partition with `Self::evalute` regardless of the
+    /// window defined in the `OVER` clause)
+    ///
+    /// ```sql
+    /// lag(x, 1) OVER (ROWS BETWEEN 2 PRECEDING AND 3 FOLLOWING)
```
Comment (comphead) [inline datafusion/physical-expr/src/window/partition_evaluator.rs:159]: interesting that lag doesn't require explicit ordering, but it works

Comment (alamb) [reply datafusion/physical-expr/src/window/partition_evaluator.rs:175]: Hmm I think the example would be better with an ORDER BY -- I will update it

**Pull request merged**